### DISCUSSION TITLE
AST of validations/matches as debug info to investigate false positives in validation

### DISCUSSION THREAD

**Discussion by zokrezyl** [json-schema-org/community#788] [Q&A]
UPDATED

Hi everyone,

I've recently been using JSON Schema and its validation mechanisms in Python for language modeling. During this process, I noticed that there doesn't seem to be any tooling or implementation that can help build an Abstract Syntax Tree (AST)-like construct from the matched input instance.

Having such a tool would be incredibly valuable because I've observed that, especially with complex or recursive schemas, it's possible to have incorrect validations (false positives) that are silently accepted. This can lead to significant issues, particularly when the validation doesn't align with the intended logic.

I'm not sure if anyone else has experimented with this idea, but I'd like to bring it to the community's attention. I'm also interested in contributing to the implementation or standardization of such an extension or approach.

To clarify, let's consider an example where I'm implementing a templating language similar to [JSON Logic](https://jsonlogic.com/) I might write a schema like the following (note that this is incomplete, with irrelevant fields omitted):

```json

{
  "$ref": "#/definitions/MyLang",
  "definitions": {
    "MyLang": {
      "oneOf": [
        { "$ref": "#/definitions/And" },
        { "$ref": "#/definitions/Any" }
      ]
    },
    "And": {
      "additionalProperties": false,
      "required": ["and"],
      "properties": {
        "and": {
          "type": "array",
          "items": {
            "oneOf": [
              { "type": "string" },
              { "type": "boolean" }
            ]
          }
        }
      }
    },
    "Any": {

        "oneOf": [
          { "type": "string" },
          {
            "type": "object",
            "not": {
              "required": ["and"]
            }
          }
        ]
      
    }
  }
}
```
Now, consider an input like this:

```json

{
  " and": [true, false]  // Note the space before "and"
}
```
In this case, the input is silently validated as correct under the schema, even though there is an unintended space before "and". While the validator might consider this a true positive according to the schema's rules, it is a false positive in the context of my intended logic.

If I could obtain the validation result in the form of an AST, I could more easily debug and determine whether the "match" was a false positive. The AST would provide a clearer view of the validation path, making it easier to spot mismatches between the expected and actual input structure.

For an example like

```json
[
  {
    "and": [
      true,
      { "not": [true] }
    ]
  },
  {
    "not": [true]
  }
]
```

The json representation of the AST would look like
```json
{
  "type": "Root",
  "children": [
    {
      "type": "AndExpression",
      "children": [
        {
          "type": "BooleanLiteral",
          "value": true
        },
        {
          "type": "NotExpression",
          "children": [
            {
              "type": "BooleanLiteral",
              "value": true
            }
          ]
        }
      ]
    },
    {
      "type": "NotExpression",
      "children": [
        {
          "type": "BooleanLiteral",
          "value": true
        }
      ]
    }
  ]
}
```

or 
```
Root
│
├── AndExpression
│   ├── BooleanLiteral: true
│   └── NotExpression
│       └── BooleanLiteral: true
│
└── NotExpression
    └── BooleanLiteral: true
```

The Root node represents the entire input structure.
The AndExpression node represents the "and" operation at the first level of the array.
Each BooleanLiteral node corresponds to a true value in the input.
The NotExpression nodes represent the "not" operations.
This AST clearly shows the structure and relationships within the input, making it easier to see whether the input conforms to the intended logic. For instance, if a validation issue arises, you could trace through this tree to see where the input deviates from the expected pattern.

If I could obtain the validation result in the form of an AST like this, it would be much easier to debug and determine whether a validation outcome was truly intended or a false positive.

for instance with the schema 

```json
[
  {
    "and": [
      true,
      { "not": [true] }
    ]
  },
  {
    "  not": [true] # note, 'mistake', leading space, will create false positive, will be valid, but difficult to spot out
  }
]
```
will produce "false positive" validation.

I hope this explanation is clear, and I'd love to hear your thoughts or see if anyone is interested in exploring this concept further.

**Comment by gregsdennis** [comment]
> build an AST like construct from the matched input instance.

I'm not sure I understand what you're suggesting. Can you edit your post to include an example of what you're looking for?

**Reply by Julian** [reply]
(Construct is [this library](https://github.com/construct/construct/) -- a slightly popular one in the Python ecosystem, though not well known outside it. I'm not 100% sure that was what was being referred to, but I read the message as if it were, so my response below was what came to mind under that interpretation.)

@zokrezyl I think the answer is "a bunch of people agree that this sort of thing would be nice, but indeed no one has written such a thing in a way which is fully general". My POC was [Seep](https://github.com/Julian/Seep) but I had simply no time to evolve it into something more full featured, and it was too early in the lifetime of `jsonschema` (the Python library) and JSON Schema (the specification) even if I did have more time.

It's possible that these days there's more room to build such a tool. As you say, speccing how it should work is a necessary step.

**Reply by zokrezyl** [reply]
Construct is cool, but bit overkill for my purpose. Seep did not do much indeed, pity you did not continue your work.

**Reply by zokrezyl** [reply]
updated my initial post

**Reply by gregsdennis** [reply]
@zokrezyl have a look at the output provided by https://json-everything.net/json-schema.  It's the current proposed output format described for the upcoming spec.  I think you'll find you get a lot of what you're looking for.

(You have an error in your schema: the `type` keyword under the `Any` definition should be a string, but you have it as an object.  Just un-nest its contents and remove `type`.)

This validation gives:

```json
{
  "valid": true,
  "evaluationPath": "",
  "schemaLocation": "https://json-everything.net/2f522b61b6#",
  "instanceLocation": "",
  "details": [
    {
      "valid": true,
      "evaluationPath": "/$ref",
      "schemaLocation": "https://json-everything.net/2f522b61b6#/definitions/MyLang",
      "instanceLocation": "",
      "details": [
        {
          "valid": false,
          "evaluationPath": "/$ref/oneOf/0",
          "schemaLocation": "https://json-everything.net/2f522b61b6#/definitions/MyLang/oneOf/0",
          "instanceLocation": "",
          "details": [
            {
              "valid": false,
              "evaluationPath": "/$ref/oneOf/0/$ref",
              "schemaLocation": "https://json-everything.net/2f522b61b6#/definitions/And",
              "instanceLocation": "",
              "errors": {
                "required": "Required properties [\"and\"] are not present"
              },
              "details": [
                {
                  "valid": false,
                  "evaluationPath": "/$ref/oneOf/0/$ref/additionalProperties",
                  "schemaLocation": "https://json-everything.net/2f522b61b6#/definitions/And/additionalProperties",
                  "instanceLocation": "/ and",
                  "errors": {
                    "": "All values fail against the false schema"
                  }
                }
              ]
            }
          ]
        },
        {
          "valid": true,
          "evaluationPath": "/$ref/oneOf/1",
          "schemaLocation": "https://json-everything.net/2f522b61b6#/definitions/MyLang/oneOf/1",
          "instanceLocation": "",
          "details": [
            {
              "valid": true,
              "evaluationPath": "/$ref/oneOf/1/$ref",
              "schemaLocation": "https://json-everything.net/2f522b61b6#/definitions/Any",
              "instanceLocation": "",
              "details": [
                {
                  "valid": false,
                  "evaluationPath": "/$ref/oneOf/1/$ref/oneOf/0",
                  "schemaLocation": "https://json-everything.net/2f522b61b6#/definitions/Any/oneOf/0",
                  "instanceLocation": "",
                  "errors": {
                    "type": "Value is \"object\" but should be \"string\""
                  }
                },
                {
                  "valid": true,
                  "evaluationPath": "/$ref/oneOf/1/$ref/oneOf/1",
                  "schemaLocation": "https://json-everything.net/2f522b61b6#/definitions/Any/oneOf/1",
                  "instanceLocation": "",
                  "details": [
                    {
                      "valid": false,
                      "evaluationPath": "/$ref/oneOf/1/$ref/oneOf/1/not",
                      "schemaLocation": "https://json-everything.net/2f522b61b6#/definitions/Any/oneOf/1/not",
                      "instanceLocation": "",
                      "errors": {
                        "required": "Required properties [\"and\"] are not present"
                      }
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

This is the same representation you have in your AST examples.

**Reply by gregsdennis** [reply]
For comparison, here's the validation of your second example:

```json
{
  "valid": true,
  "evaluationPath": "",
  "schemaLocation": "https://json-everything.net/0a06d6f685#",
  "instanceLocation": "",
  "details": [
    {
      "valid": true,
      "evaluationPath": "/$ref",
      "schemaLocation": "https://json-everything.net/0a06d6f685#/definitions/MyLang",
      "instanceLocation": "",
      "details": [
        {
          "valid": true,
          "evaluationPath": "/$ref/oneOf/0",
          "schemaLocation": "https://json-everything.net/0a06d6f685#/definitions/MyLang/oneOf/0",
          "instanceLocation": "",
          "details": [
            {
              "valid": true,
              "evaluationPath": "/$ref/oneOf/0/$ref",
              "schemaLocation": "https://json-everything.net/0a06d6f685#/definitions/And",
              "instanceLocation": "",
              "annotations": {
                "properties": [],
                "additionalProperties": []
              }
            }
          ]
        },
        {
          "valid": false,
          "evaluationPath": "/$ref/oneOf/1",
          "schemaLocation": "https://json-everything.net/0a06d6f685#/definitions/MyLang/oneOf/1",
          "instanceLocation": "",
          "details": [
            {
              "valid": false,
              "evaluationPath": "/$ref/oneOf/1/$ref",
              "schemaLocation": "https://json-everything.net/0a06d6f685#/definitions/Any",
              "instanceLocation": "",
              "errors": {
                "oneOf": "Expected 1 matching subschema but found 0"
              },
              "details": [
                {
                  "valid": false,
                  "evaluationPath": "/$ref/oneOf/1/$ref/oneOf/0",
                  "schemaLocation": "https://json-everything.net/0a06d6f685#/definitions/Any/oneOf/0",
                  "instanceLocation": "",
                  "errors": {
                    "type": "Value is \"array\" but should be \"string\""
                  }
                },
                {
                  "valid": false,
                  "evaluationPath": "/$ref/oneOf/1/$ref/oneOf/1",
                  "schemaLocation": "https://json-everything.net/0a06d6f685#/definitions/Any/oneOf/1",
                  "instanceLocation": "",
                  "errors": {
                    "type": "Value is \"array\" but should be \"object\""
                  },
                  "details": [
                    {
                      "valid": true,
                      "evaluationPath": "/$ref/oneOf/1/$ref/oneOf/1/not",
                      "schemaLocation": "https://json-everything.net/0a06d6f685#/definitions/Any/oneOf/1/not",
                      "instanceLocation": ""
                    }
                  ]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

**Reply by zokrezyl** [reply]
Indeed, there was a mistake in the schema, removed the "type", the type should be oneOf.

By rewriting my post, removed the example that causes "false positive", added it also to the post.
```json
[
  {
    "and": [
      true,
      { "not": [true] }
    ]
  },
  {
    "  not": [true] # note, 'mistake', leading space, will create false positive, will be valid, but difficult to spot out
  }
]
```

the output you pasted is also produced by python based check-jsonschema. The problem is that it produces lot of noise, and still does not spot out the false positive. A simple AST like representation would help to see if the desired output is generated. Thank you anyway for your effort.

**Reply by gregsdennis** [reply]
I think I found another issue with the schema.  The `MyLang` definition is checking for either `And` or `Any`, but the data looks like it should be checking for _an array of_ `Any` or `Any`.  Here's the change:

```
"MyLang": {
  "type": "array", // new
  "items": {       // new
    "oneOf": [
      { "$ref": "#/definitions/And" },
      { "$ref": "#/definitions/Any" }
    ]
  }
}
```

> The problem is that it produces lot of noise

Have a read through [this blog post](https://json-schema.org/blog/posts/interpreting-output) which explains a bit about output and why the noise is there, especially when you have branching schemas (ones that use keywords like `anyOf` or `if`/`then`/`else`).

---

It also looks like my implementation (which powers https://json-everything.net) is having an issue outputting the proper instance location.  I'll work on that.

---

It could be valuable to have the output include instance locations which aren't evaluated.  To do that, I suggest adding an `unevaluatedProperties: true` to the schema in the `MyLang` definition:

```
"MyLang": {
  "type": "array",
  "items": {
    "oneOf": [
      { "$ref": "#/definitions/And" },
      { "$ref": "#/definitions/Any" }
    ],
    "unevaluatedProperties": true
  } 
}
```

To illustrate this, here's a simple example:

```jsonc
// schema
{
  "properties": {
    "and": true
  },
  "unevaluatedProperties": true
}

// instance
{
  "and": 42,
  "not": []
}

// output
{
  "valid": true,
  "evaluationPath": "",
  "schemaLocation": "https://json-everything.net/d2c87ccdb7#",
  "instanceLocation": "",
  "annotations": {
    "properties": [
      "and"
    ],
    "unevaluatedProperties": [
      "  not"                      // here's an annotation
    ]
  },
  "details": [
    {
      "valid": true,
      "evaluationPath": "/properties/and",
      "schemaLocation": "https://json-everything.net/d2c87ccdb7#/properties/and",
      "instanceLocation": "/and"
    },
    {
      "valid": true,
      "evaluationPath": "/unevaluatedProperties",
      "schemaLocation": "https://json-everything.net/d2c87ccdb7#/unevaluatedProperties",
      "instanceLocation": "/  not"  // here's an output node
    }
  ]
}
```

You get an annotation and an output node that `  not` was validated by `unevaluatedProperties`, which means that it wasn't picked up by `/properties/not`.

**Reply by zokrezyl** [reply]
you are right about any. But as the name says the intention was to cover also array of anything except the other types.. Did not want to make too complicated the schema examples. With unevaluated properties is a great hint. 

Indeed most of the implementations for the validators are not giving hints about the position in the file. I am working on a generic wrapper for python jsonschema that will be able to load itself the file and give location hints about the errors. Maybe this should be also part of future standards as requirement or at least as recommendation for implementers.

I completely understood the role of the "noise" (please do not take it as offence), just was thinking about a less verbose output in case of successful validation. If the validation fails, the "noise" is obviously extremely usefull

**Reply by gregsdennis** [reply]
> With unevaluated properties is a great hint.

Glad that worked for you.

> I completely understood the role of the "noise" (please do not take it as offence), just was thinking about a less verbose output in case of successful validation. If the validation fails, the "noise" is obviously extremely usefull

In most cases, it's impossible to resolve that verbosity.  Isolating which errors are the "most pertinent" is generally an impossible task because the evaluator can only know the schema; it can't know the intent of the schema. 

If you have a schema that uses an `anyOf` to say a value can be either a string or a number, and it's an array, which subschema produced the "best" error?  The evaluator can't know that, so it gives you both.

**Reply by zokrezyl** [reply]
I think the best practice is to avoid anyOf where possible, use only oneOf, I learned it in a hard way

**Comment by jdesrosiers** [comment]
I'm not sure this is what you're looking for exactly, and it's not Python, so you probably can't use it anyway, but ...

My implementation uses a [JSON AST](https://github.com/hyperjump-io/json-schema/blob/main/lib/instance.js) that includes errors and annotations that apply to each node based on validation.

**Reply by zokrezyl** [reply]
this sounds promising, but I was rather looking for a simple tool/protocol. Updated my initial post
